### PR TITLE
feat: track installed modules

### PR DESCRIPTION
**Pull request by highskore**
- adds listeners for `ModuleInstalled(uint256, address)` and `ModuleUninstalled(uint256, address)` events
- when a module is installed it stores the installed module per account in a linked list using `writeInstalledModule()`, when a module is uninstalled it is removed using `removeInstalledModule()`.
- allows querying of all currently installed modules per account instance using `getInstalledModules()`
- currently the tests are skipped for Kernel account type, because the Kernel implementation doesn't always emit an event when a module is installed

### CODE REVIEW

**Comment by highskore** [comment]
looks like they made some [changes to forge fmt](https://github.com/foundry-rs/foundry/commit/77158ccee1cc0be8b647021b868a14314676a81a#diff-d5e31cdfe74fce47c291f135c75d1dc1beca1211112f268e5dd933d833f62077 ) recently which causes the lint ci to fail, not sure if it's a good idea to push the changes within this pr (the forge fmt diff is pretty big)

**Comment by kopy-kat** [comment]
Ah ok, good shout - lets leave it unformatted. And I will format in a separate PR. Should I already review this pr or are you still working on it?

File: src/test/utils/ERC4337Helpers.sol [hunk 1]
```diff
@@ -70,6 +74,33 @@ library ERC4337Helpers {
...
+                // Remove the uninstalled module from the list of installed modules
+                for (uint256 j; j < installedModules.length; j++) {
+                    if (
+                        installedModules[j].moduleAddress == module
+                            && installedModules[j].moduleType == moduleType
+                    ) {
+                        removeInstalledModule(j, logs[i].emitter);
+                        break;
+                    }
+                }
```
Comment (kopy-kat) [inline src/test/utils/ERC4337Helpers.sol:102]: could we use `removeInstalledModule` here?

File: src/test/utils/Storage.sol [hunk 2]
```diff
@@ -133,6 +133,186 @@ function getHelper(string memory helperType) view returns (address helper) {
...
+                        INSTALLED MODULE
+//////////////////////////////////////////////////////////////*/
+
+struct InstalledModule {
+    uint256 moduleType;
+    address moduleAddress;
+}
+
+// Adds new address to the installed module linked list for the given account
+// The list is stored in storage as a linked list in the following format:
```
Comment (kopy-kat) [inline src/test/utils/Storage.sol:146]: fyi we also have a list for linked lists: https://github.com/rhinestonewtf/sentinellist (also with a flavor for packing data next to addresses which could work here: https://github.com/rhinestonewtf/sentinellist/tree/feature/packed)

File: test/Diff.t.sol [hunk 3]
```diff
@@ -138,6 +137,188 @@ contract ERC7579DifferentialModuleKitLibTest is BaseTest {
...
+        InstalledModule[] memory modules = instance.getInstalledModules();
+        AccountType env = ModuleKitHelpers.getAccountType();
+        assertTrue(modules.length == (env == AccountType.SAFE ? 3 : 2));
+        uint256 expectedIndex = env == AccountType.SAFE ? 1 : 0;
+        assertTrue(
+            modules[expectedIndex].moduleAddress == newValidator
+                && modules[expectedIndex + 1].moduleAddress == newValidator1
+                && modules[expectedIndex].moduleType == MODULE_TYPE_VALIDATOR
+                && modules[expectedIndex + 1].moduleType == MODULE_TYPE_VALIDATOR
+        );
```
Comment (kopy-kat) [inline test/Diff.t.sol:165]: it looks like a lot of this code is reused a few times - could we add an internal func here to make it more readable?

File: src/test/utils/Storage.sol (hunk 2 above)
Comment (kopy-kat) [reply src/test/utils/Storage.sol:146]: this looks good for now but will probably be more maintainable in the future with the lib so we might want to rewrite it down the line

File: src/test/utils/ERC4337Helpers.sol (hunk 1 above)
Comment (highskore) [reply src/test/utils/ERC4337Helpers.sol:102]: you mean the one from `ModuleKitHelpers`?

Comment (kopy-kat) [reply src/test/utils/ERC4337Helpers.sol:102]: yeah

Comment (highskore) [reply src/test/utils/ERC4337Helpers.sol:102]: It's needs a reference to `AccountInstance` which we don't have there atm, if we add it somehow we could also swap all of the other functions imported from `Storage` to their `ModuleKitHelpers` equivalent

Comment (kopy-kat) [reply src/test/utils/ERC4337Helpers.sol:102]: ah ok yeah we can leave it then

**Comment by kopy-kat** [comment]
this looks good to me, thanks! could you please merge from main and fix the conflicts before we merge

**Pull request merged**